lightningd/htlc_set.c

```diff
@@ -181,7 +181,7 @@ void htlc_set_add(struct lightningd *ld,
 		return;
 	}
 
-	if (amount_msat_eq(set->so_far, total_msat)) {
+	if (amount_msat_greater_eq(set->so_far, total_msat)) {
```

Comment:
This implies we violate the spec.  @joostjager really wanted it to be exact, IIRC.  This might fail against other implementations, and the spec certainly allows them to fail.

But remember, with MPP, fuzzing the amount is unnecessary (or certainly much less).

```
    - MUST ensure that the total `amount_msat` of the HTLC set which arrives at the payee
      is equal to `total_msat`.
```

Reply:
I double-checked with @t-bast on IRC, and they also said they were checking for `>=` and not `==`. And at least for some of the simpler splitter strategies MPP still leaks information (e.g., the simple divide and conquer strategy employed by lnd can still leak the amount transferred).

Reply:
Yes, you're allowed to accept it, but you're not supposed to send it!

And sure, dumb divide in half will leak, but no more than non-MPP.

Reply:
> some of the simpler splitter strategies MPP still leaks information (e.g., the simple divide and conquer strategy employed by lnd can still leak the amount transferred)

On that note we recently changed our splitting algorithm in eclair, and it's behaving quite well so far (see https://github.com/ACINQ/eclair/pull/1427). This algorithm doesn't leak total amounts like lnd's divide and conquer does and requires less initial failed attempts in practice (but costs more CPU though). I'm curious what c-lightning is doing for MPP-split at the moment, and what algorithms you're considering for a (potential) upgrade.

Reply:
I will add an opt-out flag to the amount-fuzzing in shadow routes and the `getroute` request, if used in combination with MPP. Until we have MPP implemented (I'm working on that right now) we shouldn't disable fuzzing.

The commit in question is on the receiving end, where we should indeed be more permissive than what we send out. I propose I fixup the fuzzing in the MPP PR and leave it as is here where it is still useful since we have a single part at most here.

Reply:
Thanks for sharing the MPP logic with us @t-bast, let me try to rephrase your strategy, to see if I understood it correctly.

 - You split only amounts above 15ksat, anything below that is considered small enough to go through eventually.
 - You only split into up to 6 parts
 - You compute a number of routes initially using Yen's k-shortest paths and draw from those routes
 - When trying a route you update your estimated capacities for edges in all routes in order to avoid overloading edges.
 - Amounts are randomized in the range [0.2, 1] of the estimated capacity of the smallest channel on the route

It seems pretty similar to what my code does:
 - We have two splitter phases: a `presplit` modifier which splits large amounts before even starting a first attempt, and an `adaptive_splitter` which reacts to new information about edges and failed payments (mainly capacity estimates and disabled channels)
 - The `presplit` modifier will split anything larger than our threshold (10ksat at the moment) so that's pretty much the opposite of what you're doing. This is derived from a recent round of probing I did (see results below).
 - When splitting we fuzz the amount by +/- 10% uniformly around the target split amount (10ksat in the presplit case, or 1/2 previous attempt in the adaptive splitter).
 - We track estimates and new information we learn about channels and nodes as `channel_hints` in the root of the payment tree
 - We iteratively ask for new routes using `getroute` for each attempt, this is a bit more costly than doing it in batch with Yen's algorithm, but we can provide up to date info on the channel state, and reduce our in-memory state-management during a payment.

About the capacity probes I did: I implemented a simple binary search on the payment amount and then iterated through all nodes and tested the maximum single-part payment that will reach the destination. After testing about 35% of the network the histogram of successful capacities looks like this:

![probed-channel-capacities](https://user-images.githubusercontent.com/120117/86369896-5455d080-bc7f-11ea-9d5a-03483d0d4bfe.png)

The 80th percentile is marked in black, and the 50th percentile marked in red. With 10ksat for the pre-splitter we end up with parts that are ~1USD in size, and that will succeed in 80% of cases without requiring further splits. Our goal here is to be rather aggressive, not optimize for fees, but for speed. At 10ksat the proportional fee still dominates the overall cost, so we're confident that adding a couple more parts is not going to be too costly.

Notice that these measurements use my usual worst-case assumption of any sender-receiver pair being equally likely, which we pretty much know is too pessimistic. Not surprisingly the bottlenecks on the path are clustered around the 100ksat and 500ksat sizes, which are the most popular total capacity sizes in the network today.

We also plan to keep measuring and adapting these constants as we learn new things and the network evolves.

So overall we seem to agree on the strategy, but have chosen different parameters, let's maybe open an issue discussing these strategies on https://github.com/lightningnetwork/lightning-rfc to discuss tradeoffs and experiences :-)

Reply:
Thanks for all the details! I agree that it will be great to discuss at length heuristics we're trying and how well they seem to behave on real payments, as all simulations we do will never accurately reflect the reality of the unknown remote balances and the effectiveness of retries.

Our strategies are very similar indeed. Your summary of mine is correct, except maybe one point to clarify: I run Yen's k-shortest path not with the amount requested but instead with 15k sats, and then as you understood I "fill" these routes depending the amount they're able to carry (+ some randomization). I think this makes it even closer to your strategy of pre-splitting (but with the limitation that I impose a maximum number of parts of 6 - although configurable by the user at runtime if necessary - and this upper bound is ignored when paying a direct peer, which is a very useful heuristic when coupled with trampoline).

One thing that's always been bothering me: eclair considers that you can send up to the channel's capacity through a link, I've always wondered if using instead `channel capacity / 2` instead would yield better results. I've been holding off on implementing this change as channel capacities are (hopefully) much bigger than payment amounts so it probably doesn't matter, but I'm curious to know if you've given it some thoughts.

Thanks for sharing your graph, my analysis was only based on channel capacity, not on actual success of sending payments, so my numbers may not be very accurate.

Reply:
> The commit in question is on the receiving end, where we should indeed be more permissive than what we send out. I propose I fixup the fuzzing in the MPP PR and leave it as is here where it is still useful since we have a single part at most here.

No, let's not encourage other implementations to break the spec.  If you disagree, then this needs to be debated on the lightning-rfc spec: please open an issue.

Meanwhile, taking into account our responsibility as joint custodians of the spec, I think we need to be strict (unless it's proven that we need to allow this).

Reply:
I removed the offending commit, so now we're following the RFC to the letter again :-)